### ISSUE TITLE
FormatString not raised for non final variable?

### ISSUE THREAD

**Issue by csrma** [google/error-prone#5151]
The following code (focus on `final`) is detected by the `FormatString` rule.
```
public String bugDemo() {
    final String a = "e = %++10.4f";
    return String.format(a, Math.E);
}
```
However, if we remove the `final`, and only declare `String a = "e = %++10.4f";`, then `FormatString` is not detected.

Now, to detect the final variable, it is still required to track the value of `a` to the `String.final` call, right? Then, why removing the final variable no longer works? Is it a desired behaviour or a bug?
I think if error_prone can track the value when declared as a final variable, it should be able to track whether the non-final String is also a hard-coded bad format string. It may not raise the warning if the value is returned from a method call (if inter-method data-flow analysis is not feasible). But if the hard-coded value comes from inside the method, it seems to me error_prone can detect it since it can detect the final variable version.

By the way, `InlineFormatString` is raised (correctly) on both occasions.

ErrorProne: `2.40.0`
Dataflow-ErrorProne: `3.42.0`
Java: `JDK 21`
Run Method: `CLI`

**Comment by graememorgan** [comment]
Interesting one. I love a good bug that makes me go away and read the spec. :)

ErrorProne doesn't do its own data flow analysis here, it relies on Javac tagging [constant expressions](https://docs.oracle.com/javase/specs/jls/se21/html/jls-15.html#jls-15.29). These include "Simple names ([§6.5.6.1](https://docs.oracle.com/javase/specs/jls/se21/html/jls-6.html#jls-6.5.6.1)) that refer to constant variables ([§4.12.4](https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html#jls-4.12.4))".

[§4.12.4](https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html#jls-4.12.4) notes "A constant variable is a final variable of primitive type or type String that is initialized with a constant expression (§15.29)". So I think, surprisingly, "effectively final" and "final" differ in that the former are not _constants_.

**Comment by csrma** [comment]
Thanks for looking at this so promptly.

Since the issue stems from Javac itself, hence I think it is observed throughout error_prone wherever "constant string" tracking is required. For example, I also saw this issue for the [DuplicateMapKeys](https://errorprone.info/bugpattern/DuplicateMapKeys) rule.

`DuplicateMapKeys` is detected in the following code. However, removing `final` from the `final String foo = "Foo";` declaration evades the detection.
```
public Map<String, String> showBug(){
    final String foo = "Foo";
    String bar = "Bar";
    Map<String, String> map = Map.ofEntries(
        Map.entry(foo, bar),
        Map.entry(foo, bar)
    );
    return map;
}
```
